File: sdk/ts/client/src/withAction.ts
import type { Actor, TraceLike } from "./types.js";

type Check =
  | { name: string; pass: true; note?: string; approver?: string }
  | { name: string; pass: false; note?: string };

export async function withAction<T>(opts: {
  trace: TraceLike;
  type: string;
  actor: Actor;
  target?: string;
  params?: Record<string, unknown>;
  onApproval?: (decision: { actionId: string; checks?: string[] }) => Promise<void> | void;
  evidence?: {
    onSuccess?: (result: T) => Check[];
    onError?: (err: unknown) => Check[];
  };
  run: () => Promise<T>;
}): Promise<T> {
  const { trace, type, actor, target, params, onApproval, evidence, run } = opts;
  const decision = await trace.propose({ type, actor, target, params });

  if (decision.status === "requires_approval" && onApproval) {
    await onApproval({ actionId: decision.actionId, checks: decision.checks });
  }

  try {
    const result = await run();
    const checks = evidence?.onSuccess
      ? evidence.onSuccess(result)
      : [{ name: "action_executed", pass: true as const }];
    await trace.evidence(decision.actionId, checks);
    return result;
  } catch (err) {
    const checks = evidence?.onError
      ? evidence.onError(err)
      : [{ name: "action_failed", pass: false as const, note: String((err as Error)?.message ?? err) }];
    await trace.evidence(decision.actionId, checks);
    throw err;
  }
}

File: sdk/python/trace_client/trace_client/with_action.py
from __future__ import annotations
from typing import Any, Callable, Awaitable, Dict, List, Optional
import asyncio


async def with_action(
    *,
    trace,
    type: str,
    actor: Dict[str, Any],
    target: Optional[str] = None,
    params: Optional[Dict[str, Any]] = None,
    on_approval: Optional[Callable[[Dict[str, Any]], Awaitable[None]]] = None,
    run: Callable[[], Awaitable[Any]],
    evidence_on_success: Callable[[Any], List[Dict[str, Any]]],
    evidence_on_error: Callable[[Exception], List[Dict[str, Any]]],
) -> Any:
    decision = trace.propose(type=type, actor=actor, target=target, params=params)

    if decision.get("status") == "rejected":
        raise RuntimeError(f"TRACE decision rejected for {decision.get('actionId')}")

    if decision.get("status") == "requires_approval" and on_approval:
        await on_approval({"actionId": decision["actionId"]})

    try:
        result = await run()
        checks = evidence_on_success(result)
        if not checks:
            raise RuntimeError("evidence_on_success must return at least one check")
        trace.evidence(decision["actionId"], checks)
        return result
    except Exception as err:  # noqa: BLE001
        checks = evidence_on_error(err) if evidence_on_error else []
        if checks:
            try:
                trace.evidence(decision["actionId"], checks)
            except Exception:
                pass
        raise
